tv_apps/tv_shared/js/cards/application.js

```diff
@@ -38,20 +42,46 @@
     });
   });
 
-  Application.prototype.launch = function app_launch(args) {
-    if (this.nativeApp && this.nativeApp.launch) {
-      this.nativeApp.launch(args);
-    }
-  };
-
   Application.prototype.serialize = function app_serialize() {
     return {
       manifestURL: this.nativeApp.manifestURL,
       name: this.name,
       type: 'Application',
+      thumbnail: this.thumbnail,
+      launchURL: this.launchURL,
       group: this.group
     };
   };
 
+  Application.prototype.launch = function app_launch(args) {
+    if (this.nativeApp && this.nativeApp.launch && !this.launchURL) {
+      this.nativeApp.launch(args);
+    } else {
+      if (!Application._iacPort) {
+        console.error('no iacPort found, we cannot launch Application');
+        return;
+      }
+
+      Application._iacPort.postMessage({
+        'manifestURL': this.nativeApp.manifestURL,
+        'timestamp': (new Date()).getTime(),
+        'url': this.launchURL
+      });
+    }
+  };
+
+  window.addEventListener('load', function _retrieveIACPort() {
+    window.removeEventListener('load', _retrieveIACPort);
+
+    navigator.mozApps.getSelf().onsuccess = function(evt) {
+      var app = evt.target.result;
+      if (app) {
+        app.connect('customlaunchpath').then(function onAccepted(ports) {
+          Application._iacPort = ports[0];
```

Comment:
This part is directly cut and pasted from `AppBookmark`. And I think it is because all `application`s (originally, all `app_bookmark`s should use the same one IAC port. So we put it under scope of class `Application`. But I think it also works if we put it under prototype of Application.